### PR TITLE
[Server][Bugfix] Test if the environ variable DISPLAY is defined

### PR DESCRIPTION
**Pull request by rldhont**
## Description
Avoid that DISPLAY environ variable is mandatory.

## Checklist
- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by m-kuhn** [comment]
Where does the "offscreen" value come from?
It seems not to be too portable, but I didn't look into this enough to be able to propose something better.
https://github.com/ariya/phantomjs/issues/14376#issuecomment-298971645

**Comment by rldhont** [comment]
@dmarteau can you review this PR ? thanks.

**Comment by dmarteau** [comment]
@m-kuhn 
Please have a look at http://doc.qt.io/qt-5/embedded-linux.html
Setting this variable prevent Qt looking for display environnment where it does not exists: as an example, if you try to intitialize a QgsApplication from a docker container then it will fail with the error that Qt "cannot find display".

This may depends of the device we are running from, but setting 'offscreen' use 
a platform-independent qt rendering plugin that render to offscreen and should work everywhere. 

The problem is encountered with people trying to use the QGIS3 server on platform with no DISPLAY (docker, some linux server config with no X installed).

See also https://stackoverflow.com/questions/21488072/what-is-the-use-of-various-qt-platform-plugins.

I don't know what is the best solution, but people should not struggle with this and get frustrated when they have no display defined when they use the fcgi server. This a very nasty problem which is not well documented.

**Comment by m-kuhn** [comment]
@dmarteau I was mainly wondering for the "offscreen" value which I unfortuantely couldn't find in any official Qt doc. No objections to offscreen rendering or the objectives of this PR at all.

I can't see any harm in this approach, so if it helps somewhere that's fine and if we find a better value in the future that's compatible with a wider variety of platforms, we will be able to do this.

**Comment by elpaso** [comment]
Be careful here: AFAIK the standard way on docker is to use `xvfb` and the issue at least with Qt4 was that if you don't have a display, any HTML rendering (for exampe labels) will crash Qt (and QGIS) without any useful log or explanation.
If this is not true anymore, please go ahead, but if an X (fake or real) server is still a requirement, please leave it.

**Comment by rldhont** [comment]
@elpaso for some users who doesn't use print, just standard WMS, the server X has not to be mandatory, it is the case for a customer. So X server is not a requirement for him.

**Comment by dmarteau** [comment]
@elpaso If DISPLAY is defined, nothing is changed. If DISPLAY is not defined, HTML rendering will crash anyway but at least the application will not crash.

We have experienced the problem in two ways:
A server with no X installed, DISPLAY is not defined and they dont output HTML but the server crash at startup.

In Docker: I use embedded QgsApplication in python (debian:stretch), no DISPLAY,  if I set QT_QPA_PLATFORM" to "offscreen", it works.

Nevertheless, I don't know which is the most appropriate, so I choose a from a generic renderer plugin.

**Comment by dmarteau** [comment]
@elpaso IMHO, if you have xvfb installed then you will have a valid DISPLAY defined, so the fix won't apply in this case.

**Comment by elpaso** [comment]
Ok, that makes sense. 
If you have any idea to prevent the crash when there is no DISPLAY that would be awesome, a server application must never crash (and never leak).

**Review by elpaso — APPROVED** [review]

**Comment by m-kuhn** [comment]
For the "regular user" out there, please leave some more traces of all the thoughts and caveats from this thread. In the sourcecode (what does "offscreen" do and evenmore where does it come from/is it documented) and in the logs (something like "DISPLAY not set, running in offscreen mode, printing will not be available.")

File: src/server/qgis_map_serv.cpp [hunk 1]
```diff
@@ -39,7 +41,24 @@ int fcgi_accept()
...
 {
-  QgsApplication app( argc, argv, getenv( "DISPLAY" ), QString(), QStringLiteral( "server" ) );
+  // Test if the environ variable DISPLAY is defined
+  // if it's not, the server is running in offscreen mode
+  // Qt supports using various QPA (Qt Platform Abstraction) back ends
+  // for rendering. You can specify the back end to use with the environment
+  // variable QT_QPA_PLATFORM when invoking a Qt-based application.
+  // Available platform plugins are: directfbegl, directfb, eglfs, linuxfb,
+  // minimal, minimalegl, offscreen, wayland-egl, wayland, xcb.
+  // http://doc.qt.io/qt-5/qpa.html
```
Comment (m-kuhn) [inline src/server/qgis_map_serv.cpp]: Thanks for the documentation. Where is the list from? I can't find it in the qpa doc, is it from stackoverflow?

**Review by m-kuhn — APPROVED** [review]
LGTM

File: src/server/qgis_map_serv.cpp (hunk 1 above)
Comment (rldhont) [reply src/server/qgis_map_serv.cpp]: I found it here https://www.ics.com/blog/qt-tips-and-tricks-part-1

Comment (m-kuhn) [reply src/server/qgis_map_serv.cpp]: That's the link I was expecting to have in the comment. Finding the QPA docs should be straightforward for all the devs.

Comment (rldhont) [reply src/server/qgis_map_serv.cpp]: Ok, I'll update the comment

Comment (m-kuhn) [reply src/server/qgis_map_serv.cpp]: Thanks, merge away

**Comment by gioman** [comment]
Sorry for commenting here if I will state the obvious and/or if is not the right place: it seems that checking if there is a graphical server installed is not working also on 2.18.

On 2.14 this workaround http://www.itopen.it/qgis-server-setup-notes/ is only needed on a headless server, while on 2.18 is needed even if QGIS server is installed on a (Linux) machine with a desktop environment.

Without the workaround doing a GetPrint on a composer with an element like a text box results always in a qgis server seg fault.

This also means that basically QGIS Server 2.18 is not working on Windows servers, at least for what it concerns GetPrint requests.

**Comment by rldhont** [comment]
@gioman it's not the same problem here! The issue you pointed is a changes between 2.14 and 2.18 in the way label is managed.

**Comment by gioman** [comment]
@rldhont yeah sorry, as I said above I wasn't sure it was the same thing.

**Pull request merged**